# Structural Variant Analysis 

## Merging across callers

After haplotype inheritance filtering we sequentially merged the four following callsets:
1. Sawfish - HiFi
2. Sniffles - ONT
3. PGGB - assembly
4. PAV - assembly

The merging pipeline requires python 3.10.6, it's easiest to setup a conda environment.

### Commands used to merge the different callers
```
 mkdir stripped_vcfs
 python strip_vcf.py -i merging_config.json -o stripped_vcfs
 python interval_tree_merge.py --diff_threshold 30 -i  stripped_vcfs/sawfish.strip.pedfilt.vcf stripped_vcfs/sniffles_ont.strip.pedfilt.vcf stripped_vcfs/pggb.strip.pedfilt.vcf stripped_vcfs/pav.strip.pedfilt.vcf -o merged_hg38.svs.vcf
```

## Comparison of merged callset to Ebert et al callset

To compare our merged call set with the one generated by Ebert et al. (https://doi.org/10.1126/science.abf7117), we preprocessed their structural variant VCF. NA12878 is the only sample that overlaps between the two datasets. Therefore, after downloading their call set, SVs were extracted specific to NA12878. This was done using the following commands:

```
wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/variants_freeze4_sv_insdel_alt.vcf.gz
mv variants_freeze4_sv_insdel_alt.vcf.gz freeze4.sv.alt.vcf.gz
bcftools index freeze4.sv.alt.vcf.gz
bcftools view --samples NA12878 freeze4.sv.alt.vcf.gz | bcftools view -Oz -i 'GT[*]="alt"' - > NA12878.freeze4.sv.alt.vcf.gz
bcftools index NA12878.freeze4.sv.alt.vcf.gz
```

To identify the intersection and mutually exclusive SVs between the two call sets, we use the following command:

```
python sv_intersect_ebert.py --base-vcf NA12878.merged_hg38.svs.sort.vcf.gz --query-vcf NA12878.freeze4.sv.alt.vcf.gz --sizesim-threshold 0.7 --flank-len 200
```
